Stop using CMSampleBuffer in MediaRecorderPrivateWriter

youennf opened the pull request:
#### 14becf89d9ce9dc33b74d5aa778b1dd21d162ef1
<pre>
Stop using CMSampleBuffer in MediaRecorderPrivateWriter
<a href="https://bugs.webkit.org/show_bug.cgi?id=284700">https://bugs.webkit.org/show_bug.cgi?id=284700</a>
<a href="https://rdar.apple.com/141498525">rdar://141498525</a>

Reviewed by Youenn Fablet and Eric Carlson.

CMSampleBuffer and its CMFormatDescription aren&apos;t serialisable over IPC. They also
have a fairly high memory requirement due to their complexity.

We move over to the platform-agnostic MediaSamplesBlock (CMSampleBuffer replacement)
and TrackInfo (CMFormatDescription replacement). Those two objects can be serialised
to be transmitted over IPC. They were currently in use by the WeBM SourceBufferParser,
we now use them in both the MediaRecorderPrivateEncoder and its MediaRecorderPrivateWriter.
For storage in the MediaRecorderPrivateEncoder we use Vector of UniqueRef&lt;MediaRecorderPrivateEncoder&gt;
so that they can be moved between queues with less complexity.

No change in observable behaviour, covered by existing tests.

* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::audioSamplesDescriptionChanged):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::maybeStartWriter):
(WebCore::MediaRecorderPrivateEncoder::processVideoEncoderActiveConfiguration):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::waitForMatchingAudio):
(WebCore::MediaRecorderPrivateEncoder::flushToEndSegment):
(WebCore::MediaRecorderPrivateEncoder::interleaveAndEnqueueNextFrame):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp:
(WebCore::MediaRecorderPrivateWriter::create):
(WebCore::MediaRecorderPrivateWriter::writeFrames):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
(WebCore::MediaRecorderPrivateWriterAVFObjC::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterAVFObjC::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterAVFObjC::writeFrame):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp:
(WebCore::mkvCodeIcForMediaVideoCodecId): Move method at the top as it needs to be declared before its users.
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterWebM::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebM::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterWebM::writeFrame):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h:

Canonical link: <a href="https://commits.webkit.org/287994@main">https://commits.webkit.org/287994@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60e1a492fc0a6337f51e62848e4bf154b9683bf6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81648 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1173 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35597 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/86187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/32638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83754 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1204 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8992 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/86187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/32638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84718 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/854 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/74298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/86187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/754 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28475 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31093 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29071 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87627 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8885 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72025 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9068 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70123 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71260 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17734 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15340 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8837 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8677 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12198 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10486 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -664,26 +679,17 @@ void MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame(VideoEncoder::Enco
-    m_encodedVideoFrames.append(MediaSampleAVFObjC::create(sample->get(), *m_videoTrackIndex));
-    Ref sampleObjC = m_encodedVideoFrames.last();
-    ASSERT(frame.isKeyFrame == sampleObjC->isSync());
+    m_encodedVideoFrames.append(makeUniqueRef<MediaSamplesBlock>(m_videoTrackInfo.get(), WTFMove(vector)));
```
It is not really clear to me why m_encodedVideoFrames should own `UniqueRef<MediaSamplesBlock>` instead of `MediaSamplesBlock`.

jyavenard replied:
We maintain 3 Deque which then gets copied into another for transfer over IPC. 

A UniqueRef actually gets moved; then just 8 bytes being moved after each transfer. 

An empty MediaSamplesBlock is 32 bytes and when you take one MSB from.tbe queue to put into another, it doesn't "move" as such; it performs a deep copy. 

Just nowhere as efficient as using a Deque of UniqueRef

Additionally ; it makes the move from RetainPtr<CMSampleBuffer> to UniqueRef<MediaSamplesBlock> much easier and a very straightforward mechanical code change.

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -714,19 +720,19 @@ Ref<GenericPromise> MediaRecorderPrivateEncoder::waitForMatchingAudio(const Medi
         return GenericPromise::createAndResolve(); // Nothing to sync.
 
-    RefPtr<MediaSample> videoFrame;
+    MediaSamplesBlock* videoFrame = nullptr;
```
It would be nicer to use RefPtr or CheckedPtr instead of raw pointers here.

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
```diff
@@ -155,28 +158,38 @@ - (void)assetWriter:(AVAssetWriter *)assetWriter didOutputSegmentData:(NSData *)
-        auto result = [m_audioAssetWriterInput appendSampleBuffer:downcast<MediaSampleAVFObjC>(sample).sampleBuffer()] ? Result::Success : Result::Failure;
-        if (result != Result::Success)
+        auto sample = toCMSampleBuffer(frame, m_audioDescription.get());
+        if (sample) {
```
We usually do if(error) instead of if (ok)

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp:
```diff
@@ -45,6 +46,20 @@ namespace WebCore {
+static const char* mkvCodeIcForMediaVideoCodecId(FourCC codec)
+{
+    switch (codec.value) {
+    case 'vp08': return mkvmuxer::Tracks::kVp8CodecId;
```
We usually return

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp:
```diff
@@ -125,56 +141,32 @@ std::unique_ptr<MediaRecorderPrivateWriter> MediaRecorderPrivateWriterWebM::crea
+    for (auto& block : sample) {
+        ASSERT(block.data);
+        Ref buffer = block.data->makeContiguous();
+        success &= m_delegate->addFrame(buffer->span(), sample.trackID(), Seconds { block.presentationTime.toDouble() }.nanosecondsAs<uint64_t>(), block.isSync());
```
Does it make sense to call addFrame if the previous failed?

jyavenard replied:
Maybe not. But in this particular use with MediaRecorder a MediaSamplesBlock only ever contain a single MediaSampleItem

We never create a MediaSamplesBlock with more than two frames. 

Only MSE does so where we pack the audio samples in 2s buffer

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -300,6 +300,9 @@ void MediaRecorderPrivateEncoder::audioSamplesDescriptionChanged(const AudioStre
         outputDescription.mChannelsPerFrame = description.mChannelsPerFrame;
+        // Use 20ms audio packets for Opus
+        if (m_audioCodec == kAudioFormatOpus)
+            outputDescription.mFramesPerPacket = description.mSampleRate * MediaTime(20000, 1000000).toDouble();
```
```suggestion
            outputDescription.mFramesPerPacket = description.mSampleRate * Seconds::fromMilliseconds(20).value();
```

jyavenard committed:
Stop using CMSampleBuffer in MediaRecorderPrivateWriter
https://bugs.webkit.org/show_bug.cgi?id=284700
rdar://141498525

Reviewed by Youenn Fablet and Eric Carlson.

CMSampleBuffer and its CMFormatDescription aren't serialisable over IPC. They also
have a fairly high memory requirement due to their complexity.

We move over to the platform-agnostic MediaSamplesBlock (CMSampleBuffer replacement)
and TrackInfo (CMFormatDescription replacement). Those two objects can be serialised
to be transmitted over IPC. They were currently in use by the WeBM SourceBufferParser,
we now use them in both the MediaRecorderPrivateEncoder and its MediaRecorderPrivateWriter.
For storage in the MediaRecorderPrivateEncoder we use Vector of UniqueRef<MediaRecorderPrivateEncoder>
so that they can be moved between queues with less complexity.

No change in observable behaviour, covered by existing tests.

* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::audioSamplesDescriptionChanged):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::maybeStartWriter):
(WebCore::MediaRecorderPrivateEncoder::processVideoEncoderActiveConfiguration):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::waitForMatchingAudio):
(WebCore::MediaRecorderPrivateEncoder::flushToEndSegment):
(WebCore::MediaRecorderPrivateEncoder::interleaveAndEnqueueNextFrame):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp:
(WebCore::MediaRecorderPrivateWriter::create):
(WebCore::MediaRecorderPrivateWriter::writeFrames):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
(WebCore::MediaRecorderPrivateWriterAVFObjC::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterAVFObjC::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterAVFObjC::writeFrame):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp:
(WebCore::mkvCodeIcForMediaVideoCodecId): Move method at the top as it needs to be declared before its users.
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterWebM::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebM::addVideoTrack):
(WebCore::MediaRecorderPrivateWriterWebM::writeFrame):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h:

Canonical link: https://commits.webkit.org/287994@main
